### DISCUSSION TITLE
gatsby-transformer-json and multiple collections in single json file

### DISCUSSION THREAD

**Discussion by marceloverdijk** [gatsbyjs/gatsby#34623] [Help] [answered]
Hi,

I'm wondering if the `gatsby-transformer-json` plugin could help me generating a static website from a big json file.

I have a single json file containing multiple collections of data, e.g.

```
{
  "customers": [
    { "id": 1, "lastName": "James", .. },
    { "id": 1, "lastName": "Jones", .. },
  ],
  "countries": [
    { "iso": "us", "name": "Unites Stated", .. },
    { "iso": "uk", "name": "United Kingdom", .. },
  ],
  .. and more collections
}
```

could I query the data like:

```
{
  allCustomersJson {
    edges {
      node {
        id
        lastName
      }
    }
  }
}
```

and:

```
{
  allCountriesJson {
    edges {
      node {
        iso
        name
      }
    }
  }
}
```

or do I have to split up the big single json file to multiple json files first?
And if so is there some hook into Gatsby that I could use to do the split up, or what is the suggested wat to do this?

**Comment by LekoArts** [comment, marked as answer]
I think the best way forward here is to split it up into multiple files. When you try using the `object` from https://www.gatsbyjs.com/plugins/gatsby-transformer-json/#configuration-options you'll see that it contains all the data then:

```js
{
  object: {
    customers: [ [Object], [Object] ],
    countries: [ [Object], [Object] ]
  }
}
```

So the plugin can't iterate over that. Even if you do `const keys = Object.keys(object).map(k => ({ name: k }))` it won't work

**Reply by marceloverdijk** [reply]
Thx @LekoArts,  yes I think splitting up is the only way at the moment.

I'm wondering if the `gatsby-transformer-json` plugin could be adapted easily or that I just copy it as source for my own plugin to do something like:

```
module.exports = {
  plugins: [
    {
      resolve: `gatsby-transformer-json`,
      options: {
        // typeName: `Json`,
        collections: [ `object.continents`, `object.countries` ]
      },
    },
  ],
}
```

Similar as the `typeName` option have a `collections` options that specifies which keys in the json define the collections.
Each collection will then have it's own GraphQL field like `allCountries` and `allContinents`.

Maybe this would make more sense:

```
module.exports = {
  plugins: [
    {
      resolve: `gatsby-transformer-json`,
      options: {
        collections: [ 
          {
            element: `object.continents`,
            typeName: `continentsJson`
          },
          {
            element: `object.countries`,
            typeName: `countries`
          }
        ]
      },
    },
  ],
}
```

or a function that would return a list of collection types similar as the function for `typeName`: `({ node, object, isArray })`

I can imagine more people will have a big json with multiple types, and something like this could help to avoid splitting... wdyt?

**Reply by LekoArts** [reply]
I haven't seen this request a lot (or at all) and it's not something we'll prioritize. It greatly increases the complexity of the plugin and could be a performance footgun to do this data massaging/juggling. I'd recommend splitting it into files, optionally with the help of a node script

**Reply by marceloverdijk** [reply]
@LekoArts I did some further experiments last evening and I noticed a huge performance difference in:

1. Having 8 json files each containing an array of together 5K nodes
2. Having 8 subfolders with in each multiple single JSON files (together also 5K JSON files) 

Option 2. took 7s --> `success source and transform nodes - 7.196s`
Option 1. took almost 5m --> `success source and transform nodes - 283.586s` 

Having each node in a separate JSON file is a huge gain...

Note before I ran both I removed the `.cache` folder.

Is this performance difference expected?

**Reply by marceloverdijk** [reply]
Reading https://www.gatsbyjs.com/docs/scaling-issues/ I also tried with __Switch off type inference for `SitePage.context`__ but that didn't really help: `success source and transform nodes - 280.176s`

**Reply by LekoArts** [reply]
My guess is that the `_.isArray` codepath here is the slow part then: https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-transformer-json/src/gatsby-node.js#L31-L78

Depending on how familiar you are with profiling using Chrome DevTools you could take a look at what exactly takes so long (here's a gist with some snippets: https://gist.github.com/pieh/af6224b42dd3c746231d9df1c54a1f14).

I can take a look later, too, if you can share a public repo with example data where this shows

**Reply by marceloverdijk** [reply]
Thx @LekoArts , I will try to profile more after work... but have no experience with it ;-)
I'm using this test repo: https://github.com/marceloverdijk/my-gatsby-site

- `/data_1_multiple` uses multiple JSON files with 1 record/node per file
- `/data_2_single` uses 1 JSON file containing all records/nodes per type

which folder to use can be configured in `/gatsby-config.js`

Note I also copied the `gatsby-transformer-json` plugin locally in the project to do some additional logging.
From that I don't think the `_.isArray` itself is causing issues. That function returns quickly even for large file.
I think this is causing the performance issue:
```
    for (let i = 0, l = parsedContent.length; i < l; i++) {
      const obj = parsedContent[i]

      await transformObject(
        obj,
        createNodeId(`${node.id} [${i}] >>> JSON`),
        getType({ node, object: obj, isArray: true })
      )
    }
```
(at least for a large file with big objects).
In my case both `drivers.json` and `races.json` contains about 1K elements. Processing `drivers.json` is quick but `races.json` is slow. Each race element is a really big object.

**Reply by marceloverdijk** [reply]
`createParentChildLink({ parent: node, child: jsonNode })` 
inside `async function transformObject(obj, id, type)` seems to be causing the performance issue.
When I remove that line I go form ~280s to ~4.5s consistently...
Even the data collections show up in GraphiQL... don't know what `createParentChildLink` does, why it is slowing it down...

**Reply by marceloverdijk** [reply]
Hi @LekoArts I did some further investigation yesterday evening.

As I mentioned above in my https://github.com/marceloverdijk/my-gatsby-site repo only the `race.json` seems to be causing the performance issue. And what I can see it happens in `createParentChildLink({ parent: node, child: jsonNode })` being called..

Now that `race.json` and every `race` element itself is quite big and contains a lot of nested data.
I thought that could maybe the issue, because it's big Gatsby might have performance issue creating the schema including filters etc.
Following https://www.gatsbyjs.com/docs/scaling-issues I tried creating the schema for the `RaceJson` myself like:
(note it only has a couple of fields just for testing purpose)

```
exports.createSchemaCustomization = ({ actions }) => {
  actions.createTypes(`
    type RaceJson implements Node @dontInfer {
      id: ID!
      jsonId: Int
      year: Int
      round: Int
    }
  `)
}
```

The result is that GraphiQL now indeed has the custom schema so that works.
But `source and transform nodes - 278.476s` still takes the same amount of time so that did not help.

Is there a way to find out why `createParentChildLink` is taking so long?
The interesting part this only takes a lot of time when using `race.json` with an array; using separate `race1.json`, `race2.json` etc the performance issue is not happening..

Any ideas?

**Reply by marceloverdijk** [reply]
Note I also found this which described a similar performance issue with `createParentChildLink`: https://github.com/gatsbyjs/gatsby/issues/21447

**Comment by marceloverdijk** [comment]
I ended up creating a custom source plugin which creates Gatsby nodes from my data structure.

I'm sharing the code here as it might help others in the future.

```
const fs = require('fs')

exports.sourceNodes = async ({ actions, createContentDigest, createNodeId }, pluginOptions) => {
  const { createNode } = actions

  // Read the json file.
  const contents = fs.readFileSync(pluginOptions.path)
  let data
  try {
    data = JSON.parse(contents)
  } catch {
    throw new Error(`Unable to parse JSON: ${pluginOptions.path}`)
  }

  async function createNodes(elements, type) {
    elements.forEach((obj, index) => {
      // Construct the json node.
      const jsonNode = {
        ...obj,
        id: createNodeId(`${type}-${index}`),
        children: [],
        parent: null,
        internal: {
          type,
          contentDigest: createContentDigest(obj),
        },
      }
      // Store the original id as jsonId.
      if (obj.id) {
        jsonNode.jsonId = obj.id
      }
      // Create the node.
      createNode(jsonNode)
    })
  }

  // Create the nodes or each type.
  createNodes(data.drivers, 'Driver')
  createNodes(data.countries, 'Country')
  createNodes(data.races, 'Race')
  // etc.
}
